LamestWebserver/LamestWebserver/Core/Singleton.cs

```diff
@@ -0,0 +1,103 @@
+﻿using System;
+using System.Reflection.Emit;
+using LamestWebserver.Synchronization;
+
+namespace LamestWebserver.Core
+{
+    /// <summary>
+    /// Provides a threadsafe, auto-initializing Singleton for a given type.
+    /// </summary>
+    /// <typeparam name="T">The type of the singleton.</typeparam>
+    public class Singleton<T> : NullCheckable
+    {
+        private UsableMutexSlim _singletonLock = new UsableMutexSlim();
+
+        private bool _initialized = false;
+        private T _instance;
+
+        private Func<T> _getInstance = null;
+
+        /// <summary>
+        /// The Instance of the Singleton. The instance is automatically initalized when you first read from it.
+        /// </summary>
+        public T Instance
+        {
+            get
+            {
+                if (!_initialized) // Checked outside for higher performance in most cases but not threadsafe. Therefore checked again inside the lock.
+                {
+                    using (_singletonLock.Lock())
+                    {
+                        if (!_initialized)
+                        {
+                            _instance = _getInstance();
+                            _initialized = true;
+                        }
+                    }
+                }
+
+                return _instance;
+            }
+        }
+
+        /// <summary>
+        /// Creates a new Singleton.
+        /// </summary>
+        /// <param name="getInstanceFunction">A function to create an instance of the given type. If null will be set to default constructor of this type.</param>
+        /// <exception cref="MissingMethodException">Throws a MissingMethodException when no getInstanceFunction is given and the type does not contain a default constructor and is no ValueType.</exception>
+        public Singleton(Func<T> getInstanceFunction = null)
```

Comment:
add `bool initializeDirectly = false` that doesn't even initialize the mutex if true